### ISSUE TITLE
A rule to handle the scope polution

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#2863]
The scope polution is a problem that the extension functions have. It should be great to have a rule to detect this problem and flag this kind of functions. An example of this type of extension functions:

```kotlin
fun String.toUser() {
  Gson.fromJson(this, User::class) // this code doesn't compile, but you get the idea
}
```

This function can be handy in some scopes but it polutes the scope of `String`. If you only have this one is not a problem, but if you have multiple the problem is bigger. And if you add this in a library the problem is even bigger because you are poluting the client scope too.

This is a really difficult problem to address because the extension functions are really handy.

My first idea: A rule that have a list of classes (we should allow all classes in a package too, for example `java.lang.*`) and it will flag any `public` top level extension function over those classes. And we should thing to add a flag so the user could decide if `internal` top level extension functions are ok or not.

Problems that I see: This rule will have false positives. For example:
```kotlin
fun Any?.exhaustive() = Unit
```

Do you have any other ideas to help with the scope polution problem? Is it worth it?

**Comment by schalkms** [comment]
I think detekt really needs a `controversial` and an `experimental` ruleset. This one would go into the latter one.
All of the rules are disabled by default.
This way you can get feedback from users happily agreeing with the intention of the rule.

**Comment by arturbosch** [comment]
Sounds like a `controversial` rule for library authors.
I would also like a controversial rule set. Especially rules like `LateinitUsage` and `SpreadOperator` should be controversial as they prohibit the usage of language features.

**Comment by cortinico** [comment]
> Do you have any other ideas to help with the scope polution problem? Is it worth it?

From my point of view, it's definitely worth to have it.
I can imagine a rule that accepts a list of types (or a package) and flags extension functions applied on those types. 

> Especially rules like LateinitUsage and SpreadOperator should be controversial as they prohibit the usage of language features.

Agree here, we should probably start collecting those rules together.
